src/opm/io/eclipse/SummaryNode.cpp

```diff
@@ -0,0 +1,113 @@
+
+namespace {
+
+constexpr bool use_number(Opm::EclIO::SummaryNode::Category category) {
+    switch (category) {
+    case Opm::EclIO::SummaryNode::Category::Block:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Connection:    [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Region:        [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Segment:
+        return true;
+    case Opm::EclIO::SummaryNode::Category::Field:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Group:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Miscellaneous: [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Well:
+        return false;
+    }
+}
+
+constexpr bool use_name(Opm::EclIO::SummaryNode::Category category) {
+    switch (category) {
+    case Opm::EclIO::SummaryNode::Category::Connection:    [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Group:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Segment:       [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Well:
+        return true;
+    case Opm::EclIO::SummaryNode::Category::Block:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Field:         [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Miscellaneous: [[fallthrough]];
+    case Opm::EclIO::SummaryNode::Category::Region:
+        return false;
+    }
+}
+
+};
+
+std::string Opm::EclIO::SummaryNode::unique_key() const {
+    std::vector<std::string> key_parts { keyword } ;
+
+    if (use_name(category))
+        key_parts.emplace_back(wgname);
+
+    if (use_number(category))
+        key_parts.emplace_back(std::to_string(number));
+
+    auto compose_key = [](std::string& key, const std::string& key_part) -> std::string {
+        constexpr auto delimiter { ':' } ;
+        return key.empty() ? key_part : key + delimiter + key_part;
+    };
+
+    return std::accumulate(std::begin(key_parts), std::end(key_parts), std::string(), compose_key);
+}
+
+bool Opm::EclIO::SummaryNode::is_user_defined() const {
+    static const std::unordered_set<std::string> udq_blacklist {
+        "AUTOCOAR",
+        "AUTOREF",
+        "FULLIMP",
+        "GUIDECAL",
+        "GUIDERAT",
+        "GUPFREQ",
+        "RUNSPEC",
+        "RUNSUM",
+        "SUMMARY",
+        "SUMTHIN",
+        "SURF",
+        "SURFACT",
+        "SURFACTW",
+        "SURFADDW",
+        "SURFADS",
+        "SURFCAPD",
+        "SURFESAL",
+        "SURFNUM",
+        "SURFOPTS",
+        "SURFROCK",
+        "SURFST",
+        "SURFSTES",
+        "SURFVISC",
+        "SURFWNUM",
+    } ;
+
+    static const std::regex user_defined_regex { "[ABCFGRSW]U[A-Z]+" } ;
+
+    const bool matched     { std::regex_match(keyword, user_defined_regex) } ;
+    const bool blacklisted { udq_blacklist.find(keyword) != udq_blacklist.end() } ;
+
+    return matched && !blacklisted;
```

Comment:
Did you consider writing this as
```C++
return !blacklisted && std::regex_match(keyword, user_defined_regex);
```
?  That way, you don't incur the cost of regular expression matching for blacklisted keywords.

Reply:
I'm not sure, in this case, if that's worth it, given the domain of the function and the relative simplicity of the regular expression.